openff/recharge/esp/qcarchive.py

```diff
@@ -0,0 +1,402 @@
+
+    reordered_density = density[reverse_ao_map[:, None], reverse_ao_map]
+    return reordered_density
+
+
+@requires_package("psi4")
+def _compute_esp(
+    qc_molecule, density, esp_settings, grid
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """Computes the ESP and electric field for a particular molecule on
+    a specified grid and using the specified settings.
+
+    Parameters
+    ----------
+    qc_molecule
+        The molecule to compute the ESP / electric field of.
+    density
+        The electron density of the molecule.
+    esp_settings
+        The settings to use when computing the ESP / electric field.
+    grid
+        The grid to evaluate the ESP and electric field on.
+
+    Returns
+    -------
+        A tuple of the evaluated ESP with shape=(n_grid_points, 1) and the electric
+        field with shape=(n_grid_points, 3)
+    """
+    import psi4
+
+    psi4_molecule = psi4.geometry(qc_molecule.to_string("psi4", "angstrom"))
+    psi4_molecule.reset_point_group("c1")
+
+    psi4_wavefunction = psi4.core.RHF(
+        psi4.core.Wavefunction.build(psi4_molecule, esp_settings.basis),
+        psi4.core.SuperFunctional(),
+    )
+    psi4_wavefunction.Da().copy(psi4.core.Matrix.from_array(density))
+
+    psi4_calculator = psi4.core.ESPPropCalc(psi4_wavefunction)
+    psi4_grid = psi4.core.Matrix.from_array(grid)
+
+    esp = numpy.array(
+        psi4_calculator.compute_esp_over_grid_in_memory(psi4_grid)
+    ).reshape(-1, 1)
+
+    # TODO: enable this once an updated psi4 is released.
+    # field = numpy.array(psi4_calculator.compute_field_over_grid_in_memory(psi4_grid))
+
+    field = numpy.zeros((len(esp), 3))
+
+    return esp, field
+
+
+@requires_package("qcportal")
+@requires_package("cmiles")
+def from_qcarchive(
+    data_set_name: str,
+    subset: Optional[Iterable[str]],
+    esp_settings: ESPSettings,
+    qcfractal_address: Optional[str] = None,
+) -> List[MoleculeESPRecord]:
+    """A function to compute the ESP and electric field from a set of precomputed
+    wavefunctions stored in a particular ``Dataset`` in the MOLSSI `QCArchive
+    <https://qcarchive.molssi.org/>`_ repository.
+
+    The ESP and electric fields are currently constructed from the stored wavefunctions
+    using the Psi4 package.
+
+    Parameters
+    ----------
+    data_set_name
+        The name of the data set to compute the ESP and field for.
+    subset
+        The SMILES representations of the subset of molecules to compute the
+        ESP and field for.
+    esp_settings
+        The settings which define how to compute the ESP and field.
+    qcfractal_address
+        An optional address to the QCFractal server instance which stores the data set.
+
+    Returns
+    -------
+        The computed records.
+    """
+
+    import cmiles.utils
+    from qcelemental.models.results import WavefunctionProperties
+
+    if esp_settings.psi4_dft_grid_settings != DFTGridSettings.Default:
```

Comment:
I think you can check the grid settings used by looking up the keywords used in the compute spec, if none are specified it is default.

Reply:
Ah good call - I'll probably handle this in a future PR then!